File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/ui/StadaStationCacheViewModel.java
package de.deutschebahn.bahnhoflive.ui;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;

import java.util.HashMap;
import java.util.Map;

import de.deutschebahn.bahnhoflive.repository.StationResource;
import de.deutschebahn.bahnhoflive.repository.StationResourceProvider;

public class StadaStationCacheViewModel extends ViewModel implements StationResourceProvider {
    private final Map<String, StationResource> cache = new HashMap<>();

    @Override
    public StationResource getStationResource(String id) {
        return getStationResource(id, new StationResource(id));
    }

    @NonNull
    private StationResource getStationResource(String id, StationResource newResource) {
        StationResource stationResource = cache.get(id);

        if (stationResource == null) {
            stationResource = newResource;
            cache.put(id, stationResource);
        } else {
            stationResource.refresh();
        }

        return stationResource;
    }

}

File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/ui/StadaStationCacheViewModel.kt
/*
 * SPDX-FileCopyrightText: 2020 DB Station&Service AG <bahnhoflive-opensource@deutschebahn.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package de.deutschebahn.bahnhoflive.ui

import android.app.Application
import android.location.Location
import androidx.lifecycle.*
import de.deutschebahn.bahnhoflive.repository.StationResource
import de.deutschebahn.bahnhoflive.repository.StationResourceProvider
import de.deutschebahn.bahnhoflive.repository.stationsearch.NearbyStopPlacesResource
import de.deutschebahn.bahnhoflive.util.openhours.OpenHoursParser

open class StadaStationCacheViewModel(application: Application) : AndroidViewModel(application),
    StationResourceProvider {
    private val cache: MutableMap<String, StationResource> = HashMap()
    val locationLiveData = MutableLiveData<Location>()
    val smoothedLocationLiveData = MediatorLiveData<Location>().apply {
        addSource(locationLiveData) { location ->
            if (location == null) {
                return@addSource
            }

            val currentValue = value

            if (currentValue == null || currentValue.distanceTo(location) > 500) {
                value = location
            }
        }
    }
    val nearbyStopPlacesResourceLiveData =
        NearbyStopPlacesResource(true).let { nearbyStopPlacesResource ->
            smoothedLocationLiveData.map {
                nearbyStopPlacesResource.apply {
                    location = it
                }
            }
        }
    val nearbyStopPlacesLiveData = nearbyStopPlacesResourceLiveData.switchMap {
        it.data
    }

    val openHoursParser by lazy {
        OpenHoursParser(application, viewModelScope)
    }

    override fun getStationResource(id: String): StationResource {
        return cache.getOrPut(id) { StationResource(openHoursParser, id) }
    }

}